keymapper/injection/global_uinputs.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+# key-mapper - GUI for device specific keyboard mappings
+# Copyright (C) 2021 sezanzeb <proxima@sezanzeb.de>
+#
+# This file is part of key-mapper.
+#
+# key-mapper is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# key-mapper is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with key-mapper.  If not, see <https://www.gnu.org/licenses/>.
+
+
+import evdev
+from evdev.ecodes import EV_KEY
+
+from keymapper.utils import DEV_NAME, is_keyboard_code
+
+
+class GlobalUInputs:
+    """Manages all uinputs that are shared between all injection processes."""
+    def __init__(self):
+        self.keyboard = None
+        self.gamepad = None
+        self.mouse = None
+
+    def prepare(self):
+        """Generate uinputs.
+
+        This has to be done in the main process before injections start.
+        """
+        # Using all EV_KEY codes broke it in one installation, the use case for
+        # keyboard_output (see docstring of Context) only requires KEY_* codes here
+        # anyway and no BTN_* code.
+        # Furthermore, python-evdev modifies the ecodes.keys list to make it usable,
+        # only use KEY_* codes that are in ecodes.keys therefore.
+        keys = list(evdev.ecodes.KEY.keys() & evdev.ecodes.keys.keys())
+        self.keyboard = evdev.UInput(
+            name="key-mapper keyboard",
+            phys=DEV_NAME,
+            events={evdev.ecodes.EV_KEY: keys},
+        )
+
+    def get_appropriate_uinput(self, key):
+        """Where to inject the event to.
+
+        Or None if there is no appropriate device.
```

Comment:
This seems problematic:
1. There might be different UInput devices which can handle the same events. 
2. We need to know the mapping of the group from which the event is coming. This mapping has to include the Target device. If that is the case this method becomes way simpler: ` def get_appropriate_uinput(self, target): return devices[target]`

Reply:
> There might be different UInput devices which can handle the same events.

you are right. Both gamepads and graphics tablets use ABS_X and ABS_Y

> This mapping has to include the Target device

any ideas on how it has to be specified? Is there a default fallback if it is not specified?

Reply:
I think we should add the target information directly to the presets in json it would look something like this:
```
{
    "mapping": {
        "1,2,1": ["BTN_1", "gamepad"],
        "1,3,1": ["BTN_2", "gamepad"], 
        "1,4,1": ["KEY_A", "keyboard"]
    }
}
```
The fallback would be not mapping the event and forwarding it. The only events which have no target, are those which are not mapped.

Reply:
> Maybe we can create some script which converts existing presets.

For example there is also `migrate_path` and backwards compatibility in `split_key`. We can add a migration, yes

How is this configured in the gui?

Reply:
And how is it configured for makros? `key(KEY_A).key(BTN_1)`

Reply:
maybe `use("gamepad").key(BTN_1).key(BTN_2).key(BTN_3)`?

If an old preset is migrated, which uinput names are used. 

Is this an advanced feature that requires editing the json config?